inappmessaging/src/main/java/com/rakuten/tech/mobile/inappmessaging/runtime/data/repositories/CampaignRepository.kt

```diff
@@ -0,0 +1,208 @@
+     * Updates the [Message.isOptedOut] as true for the provided campaign.
+     */
+    fun optOutCampaign(campaign: Message): Message?
+
+    /**
+     * Decrements the number of [Message.impressionsLeft] for provided campaign Id in the repository.
+     */
+    fun decrementImpressions(id: String): Message?
+
+    /**
+     * Increments the number of [Message.impressionsLeft] for provided campaign Id in the repository.
+     */
+    fun incrementImpressions(id: String): Message?
+
+    /**
+     * Loads campaign data from user cache.
+     */
+    fun loadCachedData()
+
+    /**
+     * Clears [messages] for last user.
+     */
+    fun clearMessages()
+}
+
+@SuppressWarnings("UnnecessaryAbstractClass")
+internal abstract class CampaignRepository : CampaignRepositoryType {
+    companion object {
+        private var instance: CampaignRepository = CampaignRepositoryImpl()
+
+        @VisibleForTesting
+        internal const val IAM_USER_CACHE = "IAM_user_cache"
+        private const val TAG = "IAM_CampaignRepo"
+
+        fun instance(): CampaignRepository = instance
+    }
+
+    private class CampaignRepositoryImpl : CampaignRepository() {
+
+        // LinkedHashMap can preserve the message insertion order.
+        // Map - Key: Campaign ID, Value: Message object
+        private val messagesHashMap = LinkedHashMap<String, Message>()
+
+        init {
+            loadCachedData()
+        }
+
+        override val messages: List<Message>
+            get() {
+                return ArrayList(messagesHashMap.values)
+            }
+
+        override var lastSyncMillis: Long? = null
+            private set
+
+        @SuppressWarnings("LongMethod", "NestedBlockDepth")
+        override fun syncWith(messageList: List<Message>, timestampMillis: Long) {
+            lastSyncMillis = timestampMillis
+            val oldList = messages
+
+            messagesHashMap.clear()
+            val shouldRetainImpressionsValue = true
+            for (newCampaign in messageList) {
+                if (newCampaign.getCampaignId().isEmpty()) {
+                    continue
+                }
+
+                var updatedCampaign = newCampaign
+                val oldCampaign = oldList.firstOrNull { it.getCampaignId() == newCampaign.getCampaignId() }
+                if (oldCampaign != null) {
+                    updatedCampaign = Message.updatedMessage(
+                        updatedCampaign,
+                        asOptedOut = oldCampaign.isOptedOut ?: false
+                    )
+
+                    if (shouldRetainImpressionsValue) {
+                        var newImpressionsLeft = oldCampaign.impressionsLeft ?: oldCampaign.getMaxImpressions()
+                        val isMaxImpressionsEdited = oldCampaign.getMaxImpressions() != newCampaign.getMaxImpressions()
+                        if (isMaxImpressionsEdited) {
+                            newImpressionsLeft += newCampaign.getMaxImpressions() - oldCampaign.getMaxImpressions()
```

Comment:
There is a possibility of this being negative value. Please consider in handling.

Reply:
Updated